ompi/mca/osc/ucx/osc_ucx_passive_target.c

```diff
@@ -0,0 +1,374 @@
+
+static int start_shared(ompi_osc_ucx_module_t *module, int target) {
+    uint64_t result_value = -1;
+    ucp_ep_h ep = OSC_UCX_GET_EP(module->comm, target);
+    ucp_rkey_h rkey = (module->state_info_array)[target].rkey;
+    uint64_t remote_addr = (module->state_info_array)[target].addr;
+    ucs_status_t status;
+
+    while (true) {
+        status = ucp_atomic_fadd64(ep, 1, remote_addr, rkey, &result_value);
+        if (status != UCS_OK) {
+            return OMPI_ERROR;
+        }
+        assert(result_value >= 0);
+        if (result_value >= TARGET_LOCK_EXCLUSIVE) {
+            status = ucp_atomic_add64(ep, (-1), remote_addr, rkey);
+            if (status != UCS_OK) {
+                return OMPI_ERROR;
+            }
+        } else {
+            break;
+        }
+    }
+
+    return OMPI_SUCCESS;
+}
+
+static int end_shared(ompi_osc_ucx_module_t *module, int target) {
+    ucp_ep_h ep = OSC_UCX_GET_EP(module->comm, target);
+    ucp_rkey_h rkey = (module->state_info_array)[target].rkey;
+    uint64_t remote_addr = (module->state_info_array)[target].addr;
+    ucs_status_t status;
+
+    status = ucp_atomic_add64(ep, (-1), remote_addr, rkey);
+    if (status != UCS_OK) {
+        return OMPI_ERROR;
+    }
+
+    return OMPI_SUCCESS;
+}
+
+static int start_exclusive(ompi_osc_ucx_module_t *module, int target) {
+    uint64_t result_value = -1;
+    ucp_ep_h ep = OSC_UCX_GET_EP(module->comm, target);
+    ucp_rkey_h rkey = (module->state_info_array)[target].rkey;
+    uint64_t remote_addr = (module->state_info_array)[target].addr;
+    ucs_status_t status;
+
+    while (result_value != TARGET_LOCK_UNLOCKED) {
+        status = ucp_atomic_cswap64(ep, TARGET_LOCK_UNLOCKED,
+                                    TARGET_LOCK_EXCLUSIVE,
+                                    remote_addr, rkey, &result_value);
+        if (status != UCS_OK) {
+            return OMPI_ERROR;
+        }
+    }
+
+    return OMPI_SUCCESS;
+}
+
+static int end_exclusive(ompi_osc_ucx_module_t *module, int target) {
+    uint64_t result_value = 0;
+    ucp_ep_h ep = OSC_UCX_GET_EP(module->comm, target);
+    ucp_rkey_h rkey = (module->state_info_array)[target].rkey;
+    uint64_t remote_addr = (module->state_info_array)[target].addr;
+    ucs_status_t status;
+
+    status = ucp_atomic_swap64(ep, TARGET_LOCK_UNLOCKED,
+                               remote_addr, rkey, &result_value);
+    if (status != UCS_OK) {
+        return OMPI_ERROR;
+    }
+
+    assert(result_value == TARGET_LOCK_EXCLUSIVE);
+
+    return OMPI_SUCCESS;
+}
+
+int ompi_osc_ucx_lock(int lock_type, int target, int assert, struct ompi_win_t *win) {
+    ompi_osc_ucx_module_t *module = (ompi_osc_ucx_module_t *)win->w_osc_module;
+    ompi_osc_ucx_lock_t *lock = NULL, *item = NULL;
+    int ret = OMPI_SUCCESS;
+
+    if (module->lock_count == 0) {
+        if (module->epoch_type.access != NONE_EPOCH &&
+            module->epoch_type.access != FENCE_EPOCH) {
```

Comment:
(now L131)
I think this is the same problem with the first long discussion. When the user call Fence with no assert, MPI library cannot decide if epoch is closed or not, and the state should always be FENCE_EPOCH.